File: src/algorithm_java/leetcode/Problem448.java
package algorithm.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Problem448 {
    public List<Integer> findDisappearedNumbers1(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            if (nums[Math.abs(nums[i] - 1)] > 0) {
                nums[nums[i] - 1] = nums[nums[i] - 1] * -1;
            }
        }
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) {
                ans.add(i + 1);
            }
        }
        return ans;
    }

    public void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public List<Integer> findDisappearedNumbers(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != nums[nums[i] - 1]) {
                swap(nums, i, nums[i] - 1);
                i--;
            }
        }
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != i + 1)
                ans.add(i + 1);
        }
        return ans;
    }
}

File: src/algorithm_cpp/problem448.cpp
class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
	int n = nums.size();
	for (int i = 0; i < nums.size(); i++) {
	    int x = (nums[i] - 1) % n;
	    nums[x] += n;
	}
	vector<int> rt;
	for (int i = 0; i < nums.size(); i++) {
	    if (nums[i] <= n) {
		rt.push_back(i + 1);
	    }
	}
	return rt;
    }
};
